base_export_async/models/delay_export.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2019 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+import json
+import operator
+import base64
+from dateutil.relativedelta import relativedelta
+
+from odoo import api, fields, models, _
+from odoo.addons.queue_job.job import job
+from odoo.addons.web.controllers.main import CSVExport, ExcelExport
+from odoo.exceptions import UserError
+
+_logger = logging.getLogger(__name__)
+
+
+class DelayExport(models.Model):
+
+    _name = 'delay.export'
+    _description = 'Allow to delay the export'
+
+    user_id = fields.Many2one('res.users', string='User', index=True)
+
+    @api.model
+    def delay_export(self, data):
+        params = json.loads(data.get('data'))
+        if not self.env.user.email:
+            raise UserError(_("You must set an email address to your user."))
+        self.with_delay().export(params)
+
+    @api.model
+    @job
+    def export(self, params):
+        export_format = params.get('format')
+        raw_data = export_format != 'csv'
+
+        model_name, fields_name, ids, domain, import_compat, context = \
+            operator.itemgetter('model', 'fields', 'ids',
+                                'domain', 'import_compat', 'context')(params)
+        user = self.env['res.users'].browse([context.get('uid')])
+        if not user or not user.email:
+            raise UserError(_("The user doesn't have an email address."))
+
+        model = self.env[model_name].with_context(
+            import_compat=import_compat, **context)
+        records = model.browse(ids) or model.search(
+            domain, offset=0, limit=False, order=False)
+
+        if not model._is_an_ordinary_table():
+            fields_name = [field for field in fields_name
+                           if field['name'] != 'id']
+
+        field_names = [f['name'] for f in fields_name]
+        import_data = records.export_data(
+            field_names, raw_data).get('datas', [])
+
+        if import_compat:
+            columns_headers = field_names
+        else:
+            columns_headers = [val['label'].strip() for val in fields_name]
+
+        if export_format == 'csv':
+            csv = CSVExport()
+            result = csv.from_data(columns_headers, import_data)
+        else:
+            xls = ExcelExport()
+            result = xls.from_data(columns_headers, import_data)
```

Comment:
I'd split the part till here to `_get_file_content` or similar

Reply:
We could but I'm not sure that it is necessary. The _get_file_content will not be used somewhere else and the method seems quite clear to me as it is.

Reply:
is not just fore making it reusable, is that this big method does too many things IMO :)
and in any case, if you want to delegate file content manipulation that's the option.

Reply:
NOTE: I don't want to block you here. This can be optional if you place a nice TODO comment saying "we should split this and that... because... bla bla" ;)

Reply:
@apineux @simahawk [ROADMAP.rst](https://github.com/OCA/maintainer-tools/tree/master/template/module/readme) is the best place for this and should settle the matter :)